File: Platform/Workspace/CSharp/Allors.Workspace/Data/ContainedIn.cs
// <copyright file="ContainedIn.cs" company="Allors bv">
// Copyright (c) Allors bv. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Data
{
    using System.Collections.Generic;
    using System.Linq;
    using Allors.Protocol.Data;
    using Allors.Workspace;
    using Allors.Workspace.Meta;

    public class ContainedIn : IPropertyPredicate
    {
        public string[] Dependencies { get; set; }

        public ContainedIn(IPropertyType propertyType = null) => this.PropertyType = propertyType;

        public IPropertyType PropertyType { get; set; }

        public IExtent Extent { get; set; }

        public IEnumerable<ISessionObject> Objects { get; set; }

        public string Parameter { get; set; }

        public Predicate ToJson() =>
            new Predicate
            {
                kind = PredicateKind.ContainedIn,
                dependencies = this.Dependencies,
                propertyType = this.PropertyType?.Id,
                extent = this.Extent?.ToJson(),
                values = this.Objects?.Select(v => v.Id.ToString()).ToArray(),
                parameter = this.Parameter,
            };
    }
}

File: dotnet/platform/workspace/typescript/libs/data/system/src/containedin.ts
import { ObjectType, PropertyType } from '@allors/meta/system';
import { ISessionObject, CompositeTypes } from '@allors/domain/system';

import { ParameterizablePredicateArgs, ParameterizablePredicate } from './ParameterizablePredicate';
import { IExtent } from './IExtent';

export interface ContainedInArgs extends ParameterizablePredicateArgs, Pick<ContainedIn, 'propertyType' | 'extent' | 'objects'> {}

export class ContainedIn extends ParameterizablePredicate {
  propertyType: PropertyType;
  extent?: IExtent;
  objects?: Array<CompositeTypes>;

  constructor(propertyType: PropertyType);
  constructor(args: ContainedInArgs);
  constructor(args: ContainedInArgs | PropertyType) {
    super();

    if ((args as PropertyType).objectType) {
      this.propertyType = args as PropertyType;
    } else {
      Object.assign(this, args);
      this.propertyType = (args as ContainedInArgs).propertyType;
    }
  }

  get objectType(): ObjectType {
    return this.propertyType.objectType;
  }

  public toJSON(): any {
    return {
      kind: 'ContainedIn',
      dependencies: this.dependencies,
      propertytype: this.propertyType.id,
      parameter: this.parameter,
      extent: this.extent,
      objects: this.objects ? this.objects.map((v) => ((v as ISessionObject).id ? (v as ISessionObject).id : v)) : undefined,
    };
  }
}
